File: bench/ocaml/merkle.ml
(* Merkle tree benchmark — matches bench/merkle.march (1024 leaves, 50 rounds)
   Uses digestif SHA-256 (pure-OCaml implementation, in the march opam switch).
   Compile:
     ocamlfind ocamlopt -package digestif -linkpkg bench/ocaml/merkle.ml \
       -o /tmp/ocaml_merkle *)

let sha256_hex s =
  Digestif.SHA256.(digest_string s |> to_hex)

type node =
  | Leaf   of string               (* (hash) *)
  | Branch of string * node * node (* (hash, left, right) *)

let hash_of = function
  | Leaf h         -> h
  | Branch (h,_,_) -> h

let make_leaf v = Leaf (sha256_hex v)

let make_branch l r =
  Branch (sha256_hex (hash_of l ^ hash_of r), l, r)

let rec build_range (vals : string array) lo hi =
  if hi - lo = 1 then make_leaf vals.(lo)
  else
    let mid = (lo + hi) / 2 in
    let l   = build_range vals lo  mid in
    let r   = build_range vals mid hi  in
    make_branch l r

let rec diff_count a b =
  if hash_of a = hash_of b then 0
  else match a, b with
    | Leaf _, _   -> 1
    | _, Leaf _   -> 1
    | Branch (_, la, ra), Branch (_, lb, rb) ->
      diff_count la lb + diff_count ra rb

let run_once () =
  let vals_a = Array.init 1024 string_of_int in
  let vals_b = Array.init 1024 (fun i ->
    if i mod 8 = 0 then "modified_" ^ string_of_int i
    else string_of_int i) in
  let ta = build_range vals_a 0 1024 in
  let tb = build_range vals_b 0 1024 in
  diff_count ta tb

let () =
  let total = ref 0 in
  for _ = 1 to 50 do total := !total + run_once () done;
  Printf.printf "%d\n" !total

File: bench/elixir/merkle.exs
# Merkle tree benchmark — matches bench/merkle.march (1024 leaves, 50 rounds)
# Uses :crypto.hash(:sha256, ...) from Erlang/OTP (hardware-accelerated on most hosts).
# Expected output: 6400
defmodule MerkleBench do
  def sha256_hex(s),
    do: :crypto.hash(:sha256, s) |> Base.encode16(case: :lower)

  # Nodes are {hash, left, right} tuples; leaves have left=nil.
  def make_leaf(val),    do: {sha256_hex(val), nil, nil}
  def make_branch(l, r) do
    {hl, _, _} = l; {hr, _, _} = r
    {sha256_hex(hl <> hr), l, r}
  end

  def hash_of({h, _, _}), do: h

  def build_range(vals, lo, hi) when hi - lo == 1,
    do: make_leaf(elem(vals, lo))
  def build_range(vals, lo, hi) do
    mid = div(lo + hi, 2)
    make_branch(build_range(vals, lo, mid), build_range(vals, mid, hi))
  end

  def diff_count({ha, _, _}, {hb, _, _}) when ha == hb, do: 0
  def diff_count({_, nil, _}, _),                        do: 1
  def diff_count(_, {_, nil, _}),                        do: 1
  def diff_count({_, la, ra}, {_, lb, rb}),
    do: diff_count(la, lb) + diff_count(ra, rb)

  def run_once do
    vals_a = 0..1023 |> Enum.map(&Integer.to_string/1) |> List.to_tuple()
    vals_b =
      0..1023
      |> Enum.map(fn i ->
        if rem(i, 8) == 0, do: "modified_#{i}", else: Integer.to_string(i)
      end)
      |> List.to_tuple()
    ta = build_range(vals_a, 0, 1024)
    tb = build_range(vals_b, 0, 1024)
    diff_count(ta, tb)
  end

  def run do
    total = Enum.reduce(1..50, 0, fn _, acc -> acc + run_once() end)
    IO.puts(total)
  end
end

MerkleBench.run()
